### DISCUSSION TITLE
rotate bug

### DISCUSSION THREAD

**Discussion by doudouodong** [libvips/libvips#3658] [General]
## Questions, enhancements, tips, etc.

Please use libvips discussions for other topics.

https://github.com/libvips/libvips/discussions

## Bug report

**Describe the bug**

rotate -18 is ok 
#include <iostream>
#include <vips/vips.h>
#include <vips/vips8>
int main(int argc, char* argv[])
{
    if (VIPS_INIT(argv[0]))
    {
        return -1;
    }
    vips::VImage image = vips::VImage::new_from_file("de4d902c36c64fbb89434cd0ebcb84bc.png", 
    vips::VImage::option()->set("access", VipsAccess::VIPS_ACCESS_SEQUENTIAL));
    image = image.rotate(-18);
    image.write_to_file("out.png");
    return 0;
}

rotate -28 is not ok 
#include <iostream>
#include <vips/vips.h>
#include <vips/vips8>
int main(int argc, char* argv[])
{
    if (VIPS_INIT(argv[0]))
    {
        return -1;
    }
    vips::VImage image = vips::VImage::new_from_file("de4d902c36c64fbb89434cd0ebcb84bc.png", 
    vips::VImage::option()->set("access", VipsAccess::VIPS_ACCESS_SEQUENTIAL));
    image = image.rotate(-28);
    image.write_to_file("out.png");
    return 0;
}

(process:35482): VIPS-WARNING **: 15:39:08.794: error in tile 0 x 0

(process:35482): VIPS-WARNING **: 15:39:08.795: error in tile 0 x 16

(process:35482): VIPS-WARNING **: 15:39:08.795: error in tile 0 x 32

(process:35482): VIPS-WARNING **: 15:39:08.795: error in tile 0 x 48

(process:35482): VIPS-WARNING **: 15:39:08.795: error in tile 0 x 64

(process:35482): VIPS-WARNING **: 15:39:08.795: error in tile 0 x 80

(process:35482): VIPS-WARNING **: 15:39:08.795: error in tile 0 x 96

(process:35482): VIPS-WARNING **: 15:39:08.795: error in tile 0 x 112

(process:35482): VIPS-WARNING **: 15:39:08.795: error in tile 0 x 128
libc++abi: terminating with uncaught exception of type vips::VError: pngload: out of order read at line 624

rotate ok 

#include <iostream>
#include <vips/vips.h>
#include <vips/vips8>
int main(int argc, char* argv[])
{
    if (VIPS_INIT(argv[0]))
    {
        return -1;
    }
    vips::VImage image = vips::VImage::new_from_file("de4d902c36c64fbb89434cd0ebcb84bc.png", 
    vips::VImage::option()->set("access", VipsAccess::VIPS_ACCESS_RANDOM));
    image = image.rotate(-28);
    image.write_to_file("out.png");
    return 0;
}

de4d902c36c64fbb89434cd0ebcb84bc.png image
https://github.com/libvips/libvips/assets/47808515/0566306c-5f48-4c07-9cca-e086bf921caa

**Environment**
(please complete the following information)
Ubuntu 16.04, MacOS 10.13
Vips: 8.14.2

**Comment by lovell** [comment]
As you've seen, a rotation of 28 degrees for these dimensions creates an output that is big enough to prevent the use of VIPS_ACCESS_SEQUENTIAL for the input, and therefore VIPS_ACCESS_RANDOM is more appropriate.

If you want to keep using VIPS_ACCESS_SEQUENTIAL you could add a [vips_image_copy_memory](https://www.libvips.org/API/current/VipsImage.html#vips-image-copy-memory) operation before the rotation.

**Reply by doudouodong** [reply]
int main(int argc, char* argv[])
{
    if (VIPS_INIT(argv[0]))
    {
        return -1;
    }
    vips::VImage image = vips::VImage::new_from_file("de4d902c36c64fbb89434cd0ebcb84bc.png", 
        vips::VImage::option()->set("access", VipsAccess::VIPS_ACCESS_SEQUENTIAL)->set("memory", true));
    image = image.rotate(-28);
    image.write_to_file("out.png");
    return 0;
}

set("memory", true) also solve the problem

**Reply by jcupitt** [reply]
It would probably be better to not set sequential, eg.:

```C++
int main(int argc, char* argv[])
{
if (VIPS_INIT(argv[0]))
{
return -1;
}
vips::VImage image = vips::VImage::new_from_file("de4d902c36c64fbb89434cd0ebcb84bc.png");
image = image.rotate(-28);
image.write_to_file("out.png");
return 0;
}
```

There's a chapter in the docs explaining how SEQUENTIAL works:

https://www.libvips.org/API/current/How-it-opens-files.html
